### PR TITLE
BUG: Repeated time-series plot causes memory leak

### LINKED ISSUE #9003: Memory leak with pandas plot
SO : http://stackoverflow.com/questions/27295220/matplotlib-simple-case-memory-leak-with-pandas

Seems to be a leak memory bug with pandas.plot. It works well with plt.plot(df.index, df.test) (see below)

Simple case here :

    import sys
    import gc
    import numpy as np
    import matplotlib
    matplotlib.use('Agg')
    import matplotlib.pyplot as plt
    import pandas as pd

    pdindex = pd.date_range(start='01/01/2013', freq='15min', end='01/01/2019')
    df = pd.DataFrame({'test':np.random.normal(0,1,len(pdindex))}, index=pdindex)

    def memplot_plot(df, i):
        df.test.plot()    
        plt.title('graph' + str(i))
        plt.savefig(str(i) + '.png', dpi=144)
        plt.close() 

    for i in range(1, 100):
        print '*******************************'
        print 'i : ' + str(i)    
        print  len( gc.get_objects())
        print sys.getsizeof(gc.get_objects())
        memplot_plot(df, i)    
        gc.collect()

And the output is (memory error as of i=6):

    *******************************
    i : 1
    74682
    325680
    *******************************
    i : 2
    290627
    1190248
    *******************************
    i : 3
    506420
    2145012
    *******************************
    i : 4
    721993
    3054204
    *******************************
    i : 5
    937566
    3865524
    *******************************
    i : 6
    1153139
    4892352
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "C:\Anaconda\lib\site- packages\spyderlib\widgets\externalshell\sitecustomize.py",    line 580, in runfile
        execfile(filename, namespace)
      File "C:/PERSO/script_backtesting.py", line 124, in <module>
        memplot_plot(df, i)    
      File "C:/PERSO/script_backtesting.py", line 107, in memplot_plot
        plt.savefig(str(i) + '.png', dpi=144)
      File "C:\Anaconda\lib\site-packages\matplotlib\pyplot.py", line 576, in savefig
        res = fig.savefig(*args, **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\figure.py", line 1470, in savefig
        self.canvas.print_figure(*args, **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\backend_bases.py", line 2192, in     print_figure
        **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\backends\backend_agg.py", line 513, in  print_png
        FigureCanvasAgg.draw(self)
      File "C:\Anaconda\lib\site-packages\matplotlib\backends\backend_agg.py", line 461, in     draw
        self.figure.draw(self.renderer)
      File "C:\Anaconda\lib\site-packages\matplotlib\artist.py", line 59, in draw_wrapper
        draw(artist, renderer, *args, **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\figure.py", line 1079, in draw
        func(*args)
      File "C:\Anaconda\lib\site-packages\matplotlib\artist.py", line 59, in draw_wrapper
        draw(artist, renderer, *args, **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\axes\_base.py", line 2092, in draw
        a.draw(renderer)
      File "C:\Anaconda\lib\site-packages\matplotlib\artist.py", line 59, in draw_wrapper
        draw(artist, renderer, *args, **kwargs)
      File "C:\Anaconda\lib\site-packages\matplotlib\axis.py", line 1103, in draw
        ticks_to_draw = self._update_ticks(renderer)
      File "C:\Anaconda\lib\site-packages\matplotlib\axis.py", line 957, in _update_ticks
        tick_tups = [t for t in self.iter_ticks()]
      File "C:\Anaconda\lib\site-packages\matplotlib\axis.py", line 903, in iter_ticks
        self.major.formatter.set_locs(majorLocs)
      File "C:\Anaconda\lib\site-packages\pandas\tseries\converter.py", line 982, in  set_locs
        self._set_default_format(vmin, vmax)
      File "C:\Anaconda\lib\site-packages\pandas\tseries\converter.py", line 966, in  _set_default_format
        format = np.compress(info['maj'], info)
      File "C:\Anaconda\lib\site-packages\numpy\core\fromnumeric.py", line 1563, in  compress
       return compress(condition, axis, out)
    MemoryError

Graph of the memory for the computer from launch the script to breack and to kill the console.
![enter image description here][1]

  [1]: http://i.stack.imgur.com/Ka4gM.png

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #9003, Closes #9307. 
- Not to clutter plot functions outside of classes.
- Changed to use classmethods rather than closures for plot funcs. 
- Not cache plot function itself on `ax._plot_data`. Changed to store `str` indicates plot kind.
- Moved `tsplot` related codes to `tseries.plotting.py`.

Because the change is little big, I think it is better to wait until 0.17. Further simplification seems to be possible if we can remove `tsplot` as not public function.

CC: @qwhelan, @TomAugspurger

### CODE REVIEW

**Comment by jreback** [comment]
I moved this to 0.17.0, but can consider for 0.16.1 if warranted

**Comment by qwhelan** [comment]
@sinhrks Thanks for cleaning this up - looks good to me and I'd agree on waiting till 0.17

**Comment by anton-molyboha** [comment]
Being a n00b, not sure how to report this better: I think there is a misprint in the patch. In pandas/tseries/plotting.py line 45:
```
    freq, eries = _maybe_resample(series, ax, kwargs)
```
should be:
```
    freq, series = _maybe_resample(series, ax, kwargs)
```
Thank you for the amazing tool, and for keeping making it better!

**Comment by sinhrks** [comment]
@anton-molyboha Thanks for correction! Fixed this, and will add further tests by 0.17.

File: pandas/tools/plotting.py [hunk 1]
```diff
@@ -1734,35 +1693,36 @@ def __init__(self, data, **kwargs):
...
-                self.plt.Axes.fill_between(ax, xdata, start, y_values, **kwds)
-                self._update_prior(y)
-                return lines
+        if not 'color' in kwds:
+            kwds['color'] = lines[0].get_color()
 
-        return plotf
+        ax.fill_between(xdata, start, y_values, **kwds)
+        cls._update_stacker(ax, stacking_id, y)
+        return lines
```
Comment (tacaswell) [inline pandas/tools/plotting.py]: :+1: on returning the artists, but I suggest returning all of the artists (including the `Patch` (iirc) return by `fill_between`.

Comment (sinhrks) [inline pandas/tools/plotting.py]: @tacaswell This is only for internal use. End users get ``Axes`` (or its array) as return value based on #4020.

Comment (tacaswell) [inline pandas/tools/plotting.py]: I think returning Axes is a mistake, the plotting methods should return the artists added, will leave longer comment on the other issue.

**Comment by sinhrks** [comment]
Rebased, and added some tests for @qwhelan 's point. I think it's ready for review.

**Comment by jreback** [comment]
@sinhrks can you rebase / update?

**Comment by sinhrks** [comment]
OK, rebased.

**Comment by sinhrks** [comment]
@TomAugspurger @jreback Pls take a look when you have a time.

**Comment by TomAugspurger** [comment]
Just gave a read through. This is touching a lot of code, but I didn't see any big problems. Let's hope our tests are good enough. I'd say merge it when your ready.

**Comment by sinhrks** [comment]
Thanks for review! Merging.

**Pull request merged**